File: CONCEPTS/FlatvsOO.R
library(tibble)
library(S7)

# Your flat data
df <- tibble::tribble(
  ~GAME_ID, ~TEAM_ID, ~PLAYER_ID, ~FGM, ~FG3M, ~FTM,
      1000L,      100L,          1L,   10L,    12L,   12L,
      1000L,      100L,          2L,    4L,     4L,    7L,
      1000L,      200L,          3L,    2L,     6L,    5L,
      1000L,      200L,          4L,    8L,     2L,    7L,
      2000L,      100L,          1L,   10L,     4L,   10L,
      2000L,      100L,          2L,   11L,     5L,    4L,
      2000L,      300L,          5L,    8L,    10L,    9L,
      2000L,      300L,          6L,    7L,     6L,    3L
)

# Define S7 classes
Player <- new_class("Player",
  properties = list(
    player_id = class_integer,
    fgm = class_integer,
    fg3m = class_integer,
    ftm = class_integer
  )
)

Game <- new_class("Game",
  properties = list(
    game_id = class_integer,
    players = class_list
  )
)

Season <- new_class("Season",
  properties = list(
    games = class_list
  )
)



season <- build_season(df)

# Demonstrate hierarchical access
cat("=== Flat/Relational Access ===\n")
cat("Player 1's FGM in game 1000:", df$FGM[df$GAME_ID == 1000 & df$PLAYER_ID == 1], "\n\n")

cat("=== Hierarchical/OO Access ===\n")
cat("Player 1's FGM in game 1000:", season@games[[1]]@players[[1]]@fgm, "\n\n")

# Show structure
cat("=== Season Structure ===\n")
cat("Number of games:", length(season@games), "\n")
cat("Players in game 1:", length(season@games[[1]]@players), "\n")
cat("Players in game 2:", length(season@games[[2]]@players), "\n")

File: CONCEPTS/FlatvsOO.py
from dataclasses import dataclass
from typing import List
import pandas as pd

# Your flat data
df = pd.DataFrame({
    'GAME_ID': [1000, 1000, 1000, 1000, 2000, 2000, 2000, 2000],
    'TEAM_ID': [100, 100, 200, 200, 100, 100, 300, 300],
    'PLAYER_ID': [1, 2, 3, 4, 1, 2, 5, 6],
    'FGM': [10, 4, 2, 8, 10, 11, 8, 7],
    'FG3M': [12, 4, 6, 2, 4, 5, 10, 6],
    'FTM': [12, 7, 5, 7, 10, 4, 9, 3]
})

# Define dataclasses
@dataclass
class Player:
    player_id: int
    fgm: int
    fg3m: int
    ftm: int

@dataclass
class Game:
    game_id: int
    players: List[Player]

@dataclass
class Season:
    games: List[Game]

# Transform flat data to hierarchical structure
def build_season(df):
    games_list = []
    
    for gid in df['GAME_ID'].unique():
        game_data = df[df['GAME_ID'] == gid]
        
        players_list = [
            Player(
                player_id=int(row['PLAYER_ID']),
                fgm=int(row['FGM']),
                fg3m=int(row['FG3M']),
                ftm=int(row['FTM'])
            )
            for _, row in game_data.iterrows()
        ]
        
        games_list.append(Game(game_id=int(gid), players=players_list))
    
    return Season(games=games_list)

season = build_season(df)

# Demonstrate hierarchical access
print("=== Flat/Relational Access ===")
flat_result = df[(df['GAME_ID'] == 1000) & (df['PLAYER_ID'] == 1)]['FGM'].values[0]
print(f"Player 1's FGM in game 1000: {flat_result}\n")

print("=== Hierarchical/OO Access ===")
print(f"Player 1's FGM in game 1000: {season.games[0].players[0].fgm}\n")

# Show structure
print("=== Season Structure ===")
print(f"Number of games: {len(season.games)}")
print(f"Players in game 1: {len(season.games[0].players)}")
print(f"Players in game 2: {len(season.games[1].players)}")